networkd: add 6rd support for sit netdevs

danderson opened the pull request:
This change allows reconfiguring the `6rd-prefix` attribute of sit interfaces, which changes them from a 6to4 tunnel into a 6rd tunnel.

Tested manually by using networkd to configure 6rd connectivity with my ISP.

Fixes #10642

danderson commented:
A few questions for the reviewer(s) to consider:
 - Should I create a new "6rd" netdev Kind, instead of overloading Kind=sit ? The kernel uses the `sit` link type for both and just changes a link attribute, but networkd doesn't necessarily have to follow this.
 - I couldn't find a useful test harness for tunnel interfaces, so I just tested the change manually on my own ISP. If there's a good place to add a test, please point me to it!

yuwata commented on man/systemd.netdev.xml:
```diff
@@ -911,6 +911,13 @@
+        <term><varname>6rdPrefix=</varname></term>
+        <listitem>
+          <para>The <literal>6rdPrefix=</literal> parameter configures the tunnel as a 6rd tunnel, with the given
+          ISP-specific IPv6 prefix.</para>
```
Please specify what kind value can be specified here. I guess an IPv6 address with non-zero prefix length.

yuwata commented on man/systemd.netdev.xml:
```diff
@@ -911,6 +911,13 @@
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>6rdPrefix=</varname></term>
```
I do not like the name of this setting. What `6rd` does stand for?

yuwata replied:
It is not strong opinion, but I prefer `IPv6RapidDeploymentPrefix=`. Maybe mentioning '6rd' in the explanation below is better, something like 'This enables IPv6 Rapid Deployment, known as 6rd.'

danderson commented on src/network/netdev/tunnel.c:
```diff
@@ -617,6 +629,38 @@ int config_parse_encap_limit(const char* unit,
+
+        r = in_addr_prefix_from_string(rvalue, AF_INET6, &p, &l);
+        if (r < 0 || l == 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "6rdPrefix prefix \"%s\" is invalid, ignoring", rvalue);
```
Fixed, now reports specific error for both cases.

yuwata commented on src/network/netdev/tunnel.c:
```diff
@@ -617,6 +629,33 @@ int config_parse_encap_limit(const char* unit,
+
+        r = in_addr_prefix_from_string(rvalue, AF_INET6, &p, &l);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "6rdPrefix \"%s\" is not a valid IPv6 prefix, ignoring", rvalue);
```
Please add the error cause in the log message, that is `"..., ignoring: %m"`. So, I think the error message can be simply `"Failed to parse 6rd prefix '%s', ignoring: %m"`, as usually the error is `EINVAL`. This way, the error message is also valid for `ENOMEM` or other errors.

danderson replied:
Done.

yuwata commented on src/network/netdev/tunnel.c:
```diff
@@ -617,6 +629,33 @@ int config_parse_encap_limit(const char* unit,
+                return 0;
+        }
+        if (l == 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "6rdPrefix prefix length of \"%s\" must be greater than zero, ignoring", rvalue);
```
Please use `0` rather than `r` here.

danderson replied:
Oops, fixed!

yuwata commented on src/network/netdev/tunnel.c:
```diff
@@ -617,6 +629,33 @@ int config_parse_encap_limit(const char* unit,
         return 0;
 }
 
+CONFIG_PARSER_PROTOTYPE(config_parse_6rd_prefix) {
```
I am not sure, but we have never used the macro in the definition...

danderson replied:
Oh, sorry! I misread the source file for the original comment. I changed the definition back.

yuwata commented:
> - I couldn't find a useful test harness for tunnel interfaces, so I just tested the change manually on my own ISP. If there's a good place to add a test, please point me to it!

I am not sure, but if it is possible, please add testcases in `test/test-network/systemd-networkd-tests.py`.

danderson commented:
Thank you for the code review so far!

I'll look at `test/test-network/systemd-netword-tests.py` and see if I can implement a test, but it'll take a bit longer than the other changes :)

danderson commented:
Test added. Sorry it took me a bit more time, I had to set up a systemd-nspawn container to install+test my build without destroying my machine :)

The test just mimics how the 6rd test works. I can only check "yes, networkd created the interface", because the custom 6rd parameter is not accessible easily from the linux userspace. Hopefully that's good enough.

danderson commented:
I tried grokking the test failure, but it made no sense to me. Looks like bionic-arm64 is unhappy because it can't find the kernel cmdline logged during boot... Which I don't think can be caused by this change?

evverx commented:
I'm here kind of accidentally so I haven't taken a closer look yet. Anyway, first of all, I have to say that I really like the second commit because the fuzzers should be fed properly (I heard when they're stuffed they are so happy that they don't find any issues :-)).

> Which I don't think can be caused by this change?

This test fails occasionally on arm64 and it indeed has nothing to do with the PR. Though, I'm not sure why nobody has looked into it :-) Maybe, it's the right time.

Regarding `systemd-networkd-tests.py`, unfortunately, currently it isn't run automatically anywhere so I run it manually from time to time (mostly to make sure that CentOS won't be broken when it's turned on). The last commit seems to break several tunnel-related tests because `6rd.network` doesn't seem to be removed, which leads to `networkd` complaining that "/run/systemd/network/6rd.network:4: Tunnel is invalid, ignoring assignment: sittun99". Something like
```diff
diff --git a/test/test-network/systemd-networkd-tests.py b/test/test-network/systemd-networkd-tests.py
index 996e38725..e1ff2dad2 100755
--- a/test/test-network/systemd-networkd-tests.py
+++ b/test/test-network/systemd-networkd-tests.py
@@ -159,7 +159,7 @@ class NetworkdNetDevTests(unittest.TestCase, Utilities):
              '25-gre-tunnel.netdev', '25-gretap-tunnel.netdev', '25-vti-tunnel.netdev', '25-vti6-tunnel.netdev',
              '12-dummy.netdev', 'gre.network', 'ipip.network', 'ip6gretap.network', 'gretun.network',
              'ip6tnl.network', '25-tap.netdev', 'vti6.network', 'vti.network', 'gretap.network', 'sit.network',
-             '25-ipip-tunnel-independent.netdev', '25-wireguard.netdev']
+             '25-ipip-tunnel-independent.netdev', '25-wireguard.netdev', '6rd.network']

     def setUp(self):
         self.link_remove(self.links)
```
should make it pass.

danderson commented:
Doh, thanks for finding that! I tested in a Debian nspawn container and didn't get any errors running the test. I wonder why CentOS behaves differently, but oh well. Fixed! And also fixed the compiler warning you found earlier (missing include).

evverx commented:
Thank you!

Interestingly, I was able to reproduce that on Fedora too so it doesn't seem to be CentOS-specific. I'm still not sure why the presence of `6rd.network` prevented `networkd` from creating other interfaces (at first sight, it looks like a bug to me, but I haven't dug deeper yet).

evverx commented:
Most likely that happened because `6rd.network` took precedence over, for example, `gretun.network` where `Name=dummy98` is used in the Match section. So everything seems to be fine. Sorry for the noise.

danderson commented:
Hmm, my guess is that I didn't let the test complete 100%, I just ^C'd it after the 6rd test passed. Lesson learned :)

evverx commented:
I think it's a bug in the test because ideally it shouldn't depend on the order in which the tests are run. That is, it should probably totally wipe out the runtime directory before it starts each test case (which would also let us get rid of those lists of configs which are hard to maintain). I assume the idea behind it is that the test is supposed to preserve whatever is in /var/run/systemd/network, which is a good idea, but it could also be implemented by mounting `tmpfs` and unmounting it at the end. This way everything would be safe no matter what :-). @ssahani what do you think?

ssahani commented:
@evverx The idea behind we keep the list is  I didn't wanted to hardcore the configs in the source code. Which does not readable at all.

The tests should run in a isolated environment with no interference. 
Setup a new machine ( Let's just for saying :) )

For each test:
1. Setup test config and environment.
2. Run the test.
3. Cleanup so that other tests should not dependent on this one.

So in this order  the should not matter at all.  Is it **cleaning properly that we need to check.**

I am OK with seting up tmpfs.

poettering commented:
looks fine to me.

yuwata commented on src/network/netdev/tunnel.h:
```diff
@@ -108,3 +111,9 @@ int config_parse_tunnel_key(const char *unit, const char *filename,
+                            unsigned line, const char *section,
+                            unsigned section_line, const char *lvalue,
+                            int ltype, const char *rvalue, void *data,
+                            void *userdata);
```
> Would you like a separate PR to update the other config parsers in netdev/ to use the same macro?

Sorry, I've forgot to reply your question. I've done it in #10734.
